### PR TITLE
Notify only after 3 failures

### PR DESCRIPTION
**Pull request by TalalChaudhry**

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 1]
```diff
@@ -100,13 +100,29 @@ public void completed(AbstractBuild r) {
...
+
         if ((result == Result.ABORTED && notifier.getNotifyAborted())
                 || (result == Result.FAILURE //notify only on single failed build
                     && previousResult != Result.FAILURE
                     && notifier.getNotifyFailure())
                 || (result == Result.FAILURE //notify only on repeated failures
                     && previousResult == Result.FAILURE
                     && notifier.getNotifyRepeatedFailure())
+                || (result == Result.FAILURE //notify only after N repeated failures
+                    && count == N
```
Comment (mfrawley) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: count will only ever reach N-1 due to the for loop constraints

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 2]
```diff
@@ -140,6 +150,14 @@ public boolean getNotifyRepeatedFailure() {
         return notifyRepeatedFailure;
     }
 
+    public int getNumberOfFailuresAfterToNotify() {
```
Comment (mfrawley) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: getFailureNotificationThreshold would be better

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 3]
```diff
@@ -100,13 +100,29 @@ public void completed(AbstractBuild r) {
             previousBuild = previousBuild.getPreviousCompletedBuild();
         } while (previousBuild != null && previousBuild.getResult() == Result.ABORTED);
         Result previousResult = (previousBuild != null) ? previousBuild.getResult() : Result.SUCCESS;
+
+        AbstractBuild<?, ?> oldBuild = project.getLastBuild();
+
+        int N = notifier.getNumberOfFailuresAfterToNotify();
```
Comment (mfrawley) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: call it failureNotificationThreshold

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 4]
```diff
@@ -54,11 +54,21 @@
...
+    private int numberOfFailuresAfterToNotify;
+    private boolean notifyFailureAfterNTimess;
     private boolean includeTestSummary;
     private CommitInfoChoice commitInfoChoice;
     private boolean includeCustomMessage;
     private String customMessage;
 
+    private static String intToString(int i) {
+        if (i == -1) {
+            return ""; //$NON-NLS-1$
```
Comment (mfrawley) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: What does this comment mean?

**Review by mfrawley — CHANGES_REQUESTED** [review]

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 5]
```diff
@@ -217,6 +244,8 @@ public SlackNotifier(final String teamDomain, final String authToken, final bool
         this.notifyUnstable = notifyUnstable;
         this.notifyBackToNormal = notifyBackToNormal;
         this.notifyRepeatedFailure = notifyRepeatedFailure;
+        this.numberOfFailuresAfterToNotify = numberOfFailuresAfterToNotify;
+        this.notifyFailureAfterNTimess = notifyFailureAfterNTimess;
```
Comment (mfrawley) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: s/Timess/Times/g

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 1 above)
Comment (TalalChaudhry) [reply src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Right. thanks

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 4 above)
Comment (TalalChaudhry) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Ah sorry. I forgot to remove this. Was for some other idea that I ditched

**Review by mfrawley — APPROVED** [review]

**Comment by TalalChaudhry** [comment]
@mfrawley  @KlausTrainer  PTAL

Added feature to notifyBackToNormalAfterNFailures as well!

**Comment by TalalChaudhry** [comment]
https://github.com/Jimdo/data-documentation/issues/143

**Review by mfrawley — APPROVED** [review]
LGTM

**Pull request merged**